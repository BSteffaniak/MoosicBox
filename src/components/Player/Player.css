.player .button {
    font-family: inherit;
    font-size: inherit;
    padding: 0px;
    color: #335d92;
    background-color: rgba(68, 107, 158, 0.1);
    border-radius: 2em;
    border: 2px solid rgba(68, 107, 158, 0);
    outline: none;
    cursor: pointer;
    width: 40px;
    height: 40px;
}

.player .button:focus {
    border: 2px solid #335d92;
}

.player .button:active {
    background-color: rgba(68, 107, 158, 0.2);
}

.player {
    display: flex;
    height: 100%;
    align-items: center;
}

.player .player-now-playing {
    display: flex;
    align-items: center;
    width: 25%;
}

.player .player-now-playing-details {
    display: flex;
    flex-direction: column;
    text-align: left;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
}

.player-album-icon {
    padding: 10px 20px;
}

.dark .player-now-playing-details-title {
    overflow: hidden;
    text-overflow: ellipsis;
}

.player-now-playing-details-artist {
    margin-top: 7px;
    overflow: hidden;
    text-overflow: ellipsis;
}

.player-now-playing-details-album {
    overflow: hidden;
    text-overflow: ellipsis;
}

.dark .player-now-playing-details-artist,
.dark .player-now-playing-details-artist a {
    color: #808080;
}

.dark .player-now-playing-details-album,
.dark .player-now-playing-details-album a {
    color: #808080;
}

.player-now-playing-details-artist,
.player-now-playing-details-album {
    font-size: 14px;
}

.player-now-playing-details a {
    text-decoration: none;
}

.player-now-playing-details a:hover {
    text-decoration: underline;
}

.player .player-media-controls {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    width: 50%;
    max-width: 1000px;
}

.player .player-media-controls-seeker {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: row;
    width: 100%;
    padding-top: 5px;
    padding-bottom: 4px;
}

.player .player-media-controls-seeker-current-time {
    width: 65px;
    text-align: right;
    padding-right: 10px;
}

.player .player-media-controls-seeker-bar {
    flex: 1;
    height: 4px;
    width: 100%;
    background-color: #222222;
    border-radius: 10px;
    position: relative;
}

.player .player-media-controls-seeker-bar-progress {
    will-change: width;
    transition-property: width;
    transition-duration: 1s;
    transition-timing-function: linear;
    height: 100%;
    background-color: #ffffff;
    border-radius: 10px;
    pointer-events: none;
}

.player-media-controls-seeker-bar-progress-trigger:hover
    + .player-media-controls-seeker-bar-progress-tooltip {
    display: block;
}

.player .player-media-controls-seeker-bar-progress-tooltip {
    display: none;
    position: absolute;
    left: 0%;
    top: -34px;
    transform: translateX(-50%);
    background-color: #ffffff;
    color: black;
    padding: 1px 4px;
    border-radius: 3px;
    pointer-events: none;
}

.player .player-media-controls-seeker-bar-progress-trigger {
    position: absolute;
    left: 0;
    right: 0;
    top: 50%;
    bottom: 0;
    padding: 20px 0;
    transform: translateY(-50%);
    cursor: pointer;
}

.player .player-media-controls-seeker-total-time {
    width: 65px;
    text-align: left;
    padding-left: 10px;
}

.player .player-track-options {
    width: 25%;
}

.player .player-track-options-buttons {
    text-align: right;
    padding-right: 20px;
}

.media-button {
    margin: 5px;
}

.play-button {
    width: 17px;
    position: relative;
    top: 2px;
    left: 2px;
}

.pause-button {
    width: 17px;
    position: relative;
    top: 2px;
}

.next-track-button {
    width: 18px;
    position: relative;
    top: 2px;
    left: 1px;
}

.previous-track-button {
    width: 18px;
    transform: scaleX(-1);
    position: relative;
    top: 2px;
    right: 1px;
}

.no-transition {
    -webkit-transition: none !important;
    -moz-transition: none !important;
    -o-transition: none !important;
    transition: none !important;
}

.show-playlist-icon {
    width: 25px;
    cursor: pointer;
}

.playlist-slideout {
    display: none;
    overscroll-behavior: contain;
    transition-property: transform;
    transition-duration: 0.2s;
    background-color: #282a2b;
    position: fixed;
    right: 0;
    top: 40px;
    bottom: 120px;
    overflow: auto;
    width: 30%;
    min-width: 500px;
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    z-index: 15;
}
