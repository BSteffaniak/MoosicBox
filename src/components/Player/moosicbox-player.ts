import * as player from '~/services/player';
import type { PlaybackUpdate } from '~/services/player';
import { type Track } from '~/services/api';
import { toTime } from '~/services/formatting';

const template = `
<div class="player">
    <div class="player-media-controls-seeker-bar">
        <div class="player-media-controls-seeker-bar-progress"></div>
        <div class="player-media-controls-seeker-bar-progress-trigger"></div>
        <div class="player-media-controls-seeker-bar-progress-tooltip"></div>
    </div>
    <div class="player-controls">
        <div class="player-now-playing">
            <div class="player-album-details">
                <div class="player-album-details-icon"></div>
                <div class="player-now-playing-details">
                    <div class="player-now-playing-details-title"></div>
                    <div class="player-now-playing-details-artist"></div>
                    <div class="player-now-playing-details-album"></div>
                </div>
            </div>
        </div>
        <div class="player-media-controls">
            <div class="player-media-controls-track">
                <button class="previous-track-button media-button button">
                    <img class="previous-track-button-icon" src="/img/next-button-white.svg" alt="Previous Track" />
                </button>
                <button class="pause-button media-button button">
                    <img class="pause-button-icon" src="/img/pause-button-white.svg" alt="Pause" />
                </button>
                <button class="play-button media-button button">
                    <img class="play-button-icon" src="/img/play-button-white.svg" alt="Play" />
                </button>
                <button class="next-track-button media-button button">
                    <img class="next-track-button-icon" src="/img/next-button-white.svg" alt="Next Track" />
                </button>
                <img class="show-playback-quality-icon" src="/img/more-options-white.svg" alt="Show Playback Quality" />
            </div>
            <div class="player-media-controls-seeker">
                <span class="player-media-controls-seeker-current-time"></span>
                //
                <span class="player-media-controls-seeker-total-time"></span>
            </div>
        </div>
        <div class="player-track-options">
            <div class="player-track-options-buttons">
                <moosicbox-player-volume></moosicbox-player-volume>
                <img class="show-playback-sessions-icon" src="/img/speaker-white.svg" alt="Show Playback Sessions" />
                <img class="show-playlist-icon" src="/img/playlist-white.svg" alt="Show Playlist" />
            </div>
            <div class="player-track-options-mobile">
                <img class="mobile-playback-options" src="/img/more-options-white.svg" alt="Show Playback Options" />
                <img class="show-playlist-icon" src="/img/playlist-white.svg" alt="Show Playlist" />
            </div>
        </div>
    </div>
    <div class="player-track-options-mobile-buttons">
        <moosicbox-player-volume></moosicbox-player-volume>
        <img class="show-playback-sessions-icon" src="/img/speaker-white.svg" alt="Show Playback Sessions" />
        <img class="show-playback-quality-icon" src="/img/more-options-white.svg" alt="Show Playback Quality" />
    </div>
    <div class="playlist-slideout">
        <div class="playlist-slideout-content">
            <moosicbox-player-playlist></moosicbox-player-playlist>
        </div>
        <div class="playlist-slideout-back-to-now-playing-top">
            Back to now playing
        </div>
        <div class="playlist-slideout-back-to-now-playing-bottom">
            Back to now playing
        </div>
    </div>
</div>
`;

class MoosicBoxHTMLElement extends HTMLElement {
    constructor() {
        super();
    }

    htmlElement<T extends HTMLElement>(query: string): T | undefined {
        return this.querySelector(query) as T | undefined;
    }

    htmlElements<T extends HTMLElement>(query: string): T[] {
        return [...this.querySelectorAll(query)] as T[];
    }
}

class MoosicboxPlayerVolume extends HTMLElement {
    constructor() {
        super();
    }
}

customElements.define('moosicbox-player-volume', MoosicboxPlayerVolume);

class MoosicboxPlayer extends MoosicBoxHTMLElement {
    playbackUpdatedListenerHandle: (
        update: Omit<PlaybackUpdate, 'sessionId'>,
    ) => void = (update) => this.handlePlaybackUpdate(update);
    seekUpdatedListenerHandle: (seek: number | undefined) => void = (seek) =>
        this.handlePlaybackUpdate({ seek });
    currentTrackLengthListenerHandle: (length: number | undefined) => void = (
        length,
    ) => this.handleCurrentTrackLengthUpdate(length);
    progressAnimationFrameHandle: (ts: number) => void = (ts) =>
        this.progressAnimationFrame(ts);

    playing = false;
    applyDrag = false;
    dragging = false;

    currentTrack?: Track | undefined;
    currentTrackLength?: number | undefined;
    seekPosition?: number | undefined;
    currentSeek?: number | undefined;

    private animationStart?: number | undefined;

    progressBar: HTMLDivElement;
    seekerCurrentTime: HTMLDivElement;
    seekerTotalTime: HTMLDivElement;

    constructor() {
        super();

        this.innerHTML = template;

        this.progressBar = this.htmlElement(
            '.player-media-controls-seeker-bar-progress',
        )!;
        this.seekerCurrentTime = this.htmlElement(
            '.player-media-controls-seeker-current-time',
        )!;
        this.seekerTotalTime = this.htmlElement(
            '.player-media-controls-seeker-total-time',
        )!;
    }

    private progressAnimationFrame(ts: number): void {
        if (!this.playing) {
            this.animationStart = undefined;

            return;
        }
        if (!this.animationStart) this.animationStart = ts;

        const elapsed = ts - this.animationStart;

        const duration = this.getCurrentTrackLength();

        if (
            typeof this.currentSeek !== 'undefined' &&
            typeof duration !== 'undefined'
        ) {
            const offset = (elapsed / 1000) * (1 / duration) * 100;

            this.progressBar.style.width = `${this.getProgressBarWidth() + offset}%`;
        }

        window.requestAnimationFrame(this.progressAnimationFrameHandle);
    }

    private setPlaying(playing: boolean) {
        this.playing = playing;

        this.htmlElements('.pause-button').forEach((e) => {
            e.style.display = playing ? 'initial' : 'none';
        });

        this.htmlElements('.play-button').forEach((e) => {
            e.style.display = playing ? 'none' : 'initial';
        });

        if (playing) {
            this.startAnimation();
        }
    }

    private startAnimation() {
        window.requestAnimationFrame((ts) => {
            this.animationStart = ts;
            window.requestAnimationFrame(this.progressAnimationFrameHandle);
        });
    }

    private updateSeekBar() {
        const duration = this.getCurrentTrackLength();

        if (typeof duration === 'undefined') return;

        this.progressBar.style.width = `${this.getProgressBarWidth()}%`;
    }

    private updateSeekerTimes() {
        this.seekerCurrentTime.innerHTML = toTime(this.currentSeek ?? 0);
        this.seekerTotalTime.innerHTML = toTime(
            this.getCurrentTrackLength() ?? 0,
        );
    }

    private getSeekPosition(): number {
        const duration = this.getCurrentTrackLength();

        if (typeof duration === 'undefined') return 0;

        return Math.max(Math.min(this.seekPosition ?? 0, duration), 0);
    }

    private getCurrentSeekPosition(): number {
        if (typeof this.currentSeek === 'undefined') return 0;

        const duration = this.getCurrentTrackLength();

        if (typeof duration === 'undefined') return 0;

        return Math.max(Math.min(this.currentSeek, duration), 0);
    }

    private getCurrentTrackLength(): number | undefined {
        return this.currentTrack?.duration ?? this.currentTrackLength;
    }

    private getProgressBarWidth(): number {
        const duration = this.getCurrentTrackLength();

        if (typeof duration === 'undefined') return 0;

        if (this.applyDrag && this.dragging) {
            return (this.getSeekPosition() / duration) * 100;
        }

        return (this.getCurrentSeekPosition() / duration) * 100;
    }

    setSeek(seek: number | undefined) {
        this.currentSeek = seek;
        this.animationStart = document.timeline.currentTime as number;
        this.updateSeekerTimes();
    }

    setCurrentTrack(track: Track | undefined) {
        this.currentTrack = track;

        this.updateSeekBar();
        this.updateSeekerTimes();
    }

    async pause() {
        await player.pause();
    }

    async play() {
        await player.play();
    }

    async nextTrack() {
        await player.nextTrack();
    }

    async previousTrack() {
        await player.previousTrack();
    }

    private handlePlaybackUpdate(update: Omit<PlaybackUpdate, 'sessionId'>) {
        if (typeof update.seek !== 'undefined') {
            this.setSeek(update.seek);
        }
        if (update.play || update.playing) {
            this.setPlaying(true);
        } else if (update.playing === false) {
            this.setPlaying(false);
        }
        if (update.tracks || typeof update.position === 'number') {
            this.setCurrentTrack(
                player.playlist()![player.playlistPosition()!],
            );
        }
    }

    private handleCurrentTrackLengthUpdate(length: number | undefined) {
        this.currentTrackLength = length;

        this.updateSeekBar();
        this.updateSeekerTimes();
    }

    connectedCallback() {
        this.htmlElements('.pause-button').forEach((e) => {
            e.addEventListener('click', () => this.pause());
        });

        this.htmlElements('.play-button').forEach((e) => {
            e.addEventListener('click', () => this.play());
        });

        this.htmlElements('.next-track-button').forEach((e) => {
            e.addEventListener('click', () => this.nextTrack());
        });

        this.htmlElements('.previous-track-button').forEach((e) => {
            e.addEventListener('click', () => this.previousTrack());
        });

        player.playbackUpdatedListener.on(this.playbackUpdatedListenerHandle);
        player.onCurrentSeekChanged(this.seekUpdatedListenerHandle);
        player.onCurrentTrackLengthChanged(
            this.currentTrackLengthListenerHandle,
        );
    }

    disconnectedCallback() {
        player.playbackUpdatedListener.off(this.playbackUpdatedListenerHandle);
        player.offCurrentSeekChanged(this.seekUpdatedListenerHandle);
        player.offCurrentTrackLengthChanged(
            this.currentTrackLengthListenerHandle,
        );
    }
}

export const tagName = 'moosicbox-player';
customElements.define(tagName, MoosicboxPlayer);
