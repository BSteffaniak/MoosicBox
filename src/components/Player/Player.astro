---
import './Player.css';
import './moosicbox-player';
---

<script src="~/components/Album/album.ts"></script>

<moosicbox-player client:load="custom-elements-ssr"></moosicbox-player>

<style></style>
