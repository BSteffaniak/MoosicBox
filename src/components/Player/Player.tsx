import { createEffect, createSignal, on, onCleanup, onMount } from 'solid-js';
import './Player.css';
import {
    currentAlbum,
    currentSeek,
    currentTrack,
    currentTrackLength,
    nextTrack,
    pause,
    play,
    playing,
    previousTrack,
    seek,
    setCurrentAlbum,
    setCurrentSeek,
    setCurrentTrack,
    setCurrentTrackLength,
} from '~/services/player';
import { A } from '@solidjs/router';
import { getAlbumArtwork } from '~/services/api';
import { toTime } from '~/services/formatting';
import { isServer } from 'solid-js/web';

let mouseX: number;

function eventToSeekPosition(element: HTMLElement): number {
    const pos = element.getBoundingClientRect()!;
    const percentage = (mouseX - pos.left) / pos.width;
    return currentTrackLength() * percentage;
}

function seekTo(event: MouseEvent): void {
    seek(Math.round(eventToSeekPosition(event.target as HTMLElement)));
}

let dragStartListener: (event: MouseEvent) => void;
let dragListener: (event: MouseEvent) => void;
let dragEndListener: (event: MouseEvent) => void;

export default function Player() {
    let progressBar: HTMLDivElement | undefined;
    let progressBarTrigger: HTMLDivElement | undefined;
    let [dragging, setDragging] = createSignal(false);
    let [applyDrag, setApplyDrag] = createSignal(false);
    let [seekPosition, setSeekPosition] = createSignal(currentSeek());

    function speedyProgressTransition() {
        progressBar?.classList.add('no-transition');
        setTimeout(() => {
            progressBar?.classList.remove('no-transition');
        }, 100);
    }

    function getSeekPosition() {
        return Math.max(Math.min(seekPosition() ?? 0, currentTrackLength()), 0);
    }

    onMount(() => {
        setCurrentTrackLength(currentTrackLength());
        setCurrentAlbum(currentAlbum());
        setCurrentTrack(currentTrack());
        setCurrentSeek(currentSeek());
        speedyProgressTransition();

        if (!isServer) {
            dragStartListener = (event: MouseEvent) => {
                if (event.button === 0) {
                    progressBar?.classList.add('no-transition');
                    setDragging(true);
                    setApplyDrag(true);
                }
            };
            dragListener = (event: MouseEvent) => {
                mouseX = event.clientX;
                if (dragging()) {
                    event.preventDefault();
                    if (!applyDrag()) return;
                }
                setSeekPosition(eventToSeekPosition(progressBarTrigger!));
            };
            dragEndListener = (event: MouseEvent) => {
                if (event.button === 0 && dragging()) {
                    setDragging(false);
                    if (!applyDrag()) return;
                    setApplyDrag(false);
                    seek(Math.round(seekPosition()!));
                    progressBar?.classList.remove('no-transition');
                    event.preventDefault();
                }
            };
            progressBarTrigger!.addEventListener(
                'mousedown',
                dragStartListener,
            );
            window.addEventListener('mousemove', dragListener);
            window.addEventListener('mouseup', dragEndListener);
        }
    });

    onCleanup(() => {
        if (!isServer) {
            progressBarTrigger!.removeEventListener(
                'mousedown',
                dragStartListener,
            );
            window.removeEventListener('mousemove', dragListener);
            window.removeEventListener('mouseup', dragEndListener);
        }
    });

    createEffect(
        on(
            () => currentSeek(),
            (newSeek, oldSeek) => {
                if (
                    typeof newSeek === 'undefined' ||
                    typeof oldSeek === 'undefined' ||
                    Math.abs(newSeek - oldSeek) > 1.2
                ) {
                    speedyProgressTransition();
                }
            },
        ),
    );

    createEffect(
        on(
            () => currentTrackLength(),
            () => {
                setSeekPosition(eventToSeekPosition(progressBarTrigger!));
            },
        ),
    );

    createEffect(
        on(
            () => playing(),
            () => {
                if (dragging()) {
                    setApplyDrag(false);
                    progressBar?.classList.remove('no-transition');
                }
            },
        ),
    );

    return (
        <div class="player">
            <div class="player-now-playing">
                <A href={`/albums/${currentTrack()?.albumId}`}>
                    <img
                        class="player-album-icon"
                        style={{ width: '70px', height: '70px' }}
                        src={getAlbumArtwork(currentTrack())}
                    />
                </A>
                <div class="player-now-playing-details">
                    <div class="player-now-playing-details-title">
                        <A href={`/albums/${currentTrack()?.albumId}`}>
                            {currentTrack()?.title}
                        </A>
                    </div>
                    <div class="player-now-playing-details-artist">
                        <A href={`/artists/${currentTrack()?.artistId}`}>
                            {currentTrack()?.artist}
                        </A>
                    </div>
                    <div class="player-now-playing-details-album">
                        Playing from:{' '}
                        <A href={`/albums/${currentTrack()?.albumId}`}>
                            {currentTrack()?.album}
                        </A>
                    </div>
                </div>
            </div>
            <div class="player-media-controls">
                <div class="player-media-controls-track">
                    <button
                        class="media-button button"
                        onClick={() => previousTrack()}
                    >
                        <img
                            class="previous-track-button"
                            src="/img/next-button-white.svg"
                            alt="Previous Track"
                        />
                    </button>
                    {playing() ? (
                        <button
                            class="media-button button"
                            onClick={() => pause()}
                        >
                            <img
                                class="pause-button"
                                src="/img/pause-button-white.svg"
                                alt="Pause"
                            />
                        </button>
                    ) : (
                        <button
                            class="media-button button"
                            onClick={() => play()}
                        >
                            <img
                                class="play-button"
                                src="/img/play-button-white.svg"
                                alt="Play"
                            />
                        </button>
                    )}
                    <button
                        class="media-button button"
                        onClick={() => nextTrack()}
                    >
                        <img
                            class="next-track-button"
                            src="/img/next-button-white.svg"
                            alt="Next Track"
                        />
                    </button>
                </div>
                <div class="player-media-controls-seeker">
                    <div class="player-media-controls-seeker-current-time">
                        {toTime(currentSeek() ?? 0)}
                    </div>
                    <div class="player-media-controls-seeker-bar">
                        <div
                            ref={progressBar}
                            class="player-media-controls-seeker-bar-progress"
                            style={{
                                width: `${Math.min(
                                    ((applyDrag() && dragging()
                                        ? getSeekPosition()!
                                        : (currentSeek() ?? 0) +
                                          (currentSeek() ?? 0) /
                                              currentTrackLength()) /
                                        currentTrackLength()) *
                                        100,
                                    100,
                                )}%`,
                            }}
                        ></div>
                        <div
                            ref={progressBarTrigger}
                            class="player-media-controls-seeker-bar-progress-trigger"
                            onClick={(e) => seekTo(e)}
                        ></div>
                        <div
                            class="player-media-controls-seeker-bar-progress-tooltip"
                            style={{
                                left: `${
                                    (getSeekPosition() / currentTrackLength()) *
                                    100
                                }%`,
                            }}
                        >
                            {toTime(Math.round(getSeekPosition()))}
                        </div>
                    </div>
                    <div class="player-media-controls-seeker-total-time">
                        {toTime(currentTrackLength())}
                    </div>
                </div>
            </div>
            <div class="player-track-options"></div>
        </div>
    );
}
