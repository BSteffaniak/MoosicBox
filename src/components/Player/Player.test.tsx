import { render } from '@solidjs/testing-library';
import Player from './Player';

describe('<Player />', () => {
    it('renders Connecting...', () => {
        const { container, unmount } = render(() => <Player />);
        expect(container).toMatchSnapshot();
        unmount();
    });
});
