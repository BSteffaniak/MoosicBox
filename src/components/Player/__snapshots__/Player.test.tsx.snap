// Vitest Snapshot v1

exports[`<Player /> > renders buttons`] = `
<div>
  <button
    class="button"
  >
    Play
  </button>
  <button
    class="button"
  >
    Stop player
  </button>
  <button
    class="button"
  >
    Start player
  </button>
  <button
    class="button"
  >
    Restart player
  </button>
</div>
`;
